File: string/20.cpp
/*
 * @Author: httermin
 * @Date: 2019-09-09 23:37:52
 */
#include "string"
#include<iostream>
#include<stack>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        int length = s.size();
        if(length%2 != 0) return false;
        stack<char> parent;
        for(int i=0;i<s.size();i++){
            switch(s[i]){
            case '(':parent.push(s[i]);break;
            case '[':parent.push(s[i]);break;
            case '{':parent.push(s[i]);break;
            case ')':if(parent.empty() || parent.top()!='(') return false; else parent.pop();break;
            case ']':if(parent.empty() || parent.top()!='[') return false; else parent.pop();break;
            case '}':if(parent.empty() || parent.top()!='{') return false; else parent.pop();break;
            default:;
            }
        }
        return parent.empty();
    }
};


int main(){
    string s = "{}[]([{}])";
    stack<char> parent;
    for (int i=0; i<s.size(); i++){
        switch(s[i]){
            case '(':
            case '[':
            case '{':parent.push(s[i]);break;
        }
    }

    Solution ss;
    cout<<ss.isValid(s)<<endl;
    return -1;
}

File: stack/20.py
# coding=UTF-8
'''
@Author: httermin
@Date: 2019-12-30 15:37:03
'''
import sys
import os

sys.path.append("..")
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2:
            return False

        stack = []

        for i in range(len(s)):
            if s[i] in ["(", "[", "{"]:
                stack.append(s[i])
            elif s[i] == ")" and len(stack) and stack[-1] == "(":
                stack.pop()
            elif s[i] == "]" and len(stack) and stack[-1] == "[":
                stack.pop()
            elif s[i] == "}" and len(stack) and stack[-1] == "{":
                stack.pop()
            else:
                return False
        if len(stack) != 0:
            return False
        return True

if __name__ == "__main__":
    a = "){"

    so = Solution()

    print(so.isValid(a))